adapt to latest theia 1.24

alxflam opened the pull request:
Hi,

with the current Theia 1.24 (and already with 1.23.0) the tree editor cannot be saved anymore.
See changes in [Theia#10736](https://github.com/eclipse-theia/theia/pull/10736) and https://github.com/alxflam/theia/commit/e62fd33ce46055d1633d781c08bf5852535f53c5.
Monaco now defines different save preferences (the keys as well as the allowed values have changed).

Changes:
- updated Theia dependency to ^1.24.0
- implement SaveableSource and the methods createSnapshot and revert (from Saveable, dummy implementations like they were already present in a previous release) as they are now required (see WorkspaceSaveResourceService#save which calls workspaceFrontendContribution#canBeSavedAs)
- the dummy implementations throw an exception mentioning that subclasses should override these methods
- adapt autoSave due to changed type and preference key
- typo fixes in documentation

I had to add a "eslint-disable-next-line import/no-deprecated" to the import of createTreeContainer as there are now two such functions, whose only difference is the type of an optional parameter (and one of these functions is deprecated). Don't know how i can trick typescript to import the not-deprecated function if they are actually named identically.

I did not implement 'onFocusChange' / 'onWindowChange', instead i kept it simple and always save with a delay (delay from the preferences). We may need to look into this once more.

Greetings, 
Alex

lucas-koehler commented on theia-tree-editor/src/browser/tree-editor-widget.tsx:
```diff
@@ -80,6 +81,18 @@ export abstract class BaseTreeEditorWidget extends BaseWidget implements Saveabl
+        return this;
+    }
+
+    createSnapshot(): object {
```
Is there a reasonable default implementation we could provide here or in the `ResourceTreeEditorWidget`? (Same for `revert`)?

lucas-koehler reviewed (commented):
Hello Alex,
thank you very much for your contribution :)
It looks pretty good to me. I'm just wondering whether we could provide a (simple) default implementation for `createSnapshot` and `revert` either in the `BaseTreeEditorWidget` or the `ResourceTreeEditorWidget`. Do you have an idea what would be needed for this?

alxflam commented:
@lucas-koehler thanks for your feedback, i'll try to summarize my observations:

- `createSnapshot` is called when 'Save As' is triggered: the snapshot returned is then usually simply the editor content (for plain monaco editors simply a string) and is then applied to the new file using `applySnapshot` (see `copyAndSave` in `WorkspaceFrontendContribution`)
- `revert` is called upon closing dirty editors, if the user chooses to discard the changes

Now `BaseTreeEditorWidget` only holds the current state in the `instanceData` field, but it cannot load the currently persisted state again (would be required for `revert`). But `createSnapshot` could simply return the current `instanceData`. The same is true for `NavigatableTreeEditorWidget`.

`ResourceTreeEditorWidget` instead could implement `revert` by calling `load` to restore the currently persisted state of the resource and in `createSnapshot` return `this.data`.

A few remarks:
- Somehow `instanceData` of `BaseTreeEditorWidget` is not used consistently: While the coffee editor example sets the member (extending from `NavigatableTreeEditorWidget`), the `ResourceTreeEditorWidget` instead uses a custom `data` member.
- When using a model server, the usual implementation of revert would be to call `undo` on the `ModelServerClient` - but there's no base class yet for model sever trees, hence no way to provide this in a generic way

For now, i have added both default implementations to `ResourceTreeEditorWidget` and only `createSnapshot` for `BaseTreeEditorWidget`. I think the `revert` for now needs to be handled by the implementing subclass. Additionally, `ResourceTreeEditorWidget` now also reuses the parents `instanceData` member.

lucas-koehler commented on theia-tree-editor/src/browser/tree-editor-widget.tsx:
```diff
@@ -79,6 +80,17 @@ export abstract class BaseTreeEditorWidget extends BaseWidget implements Saveabl
+    }
+
+    createSnapshot(): object {
+        return this.instanceData;
```
As this snapshot can be used in any way not known to us and even applied to new widgets (like in copyAndSave), I think we should not just return the object: Editing this snapshot outside the editor would then also edit the data in the editor.

I think we have two solutions:

1. deep clone the object (e.g. by using lodash cloneDeep)
2. Use a text-based snapshot approach. We could `JSON.stringify` the data and return the string. Then we may also be able to implement `applySnapshot` by parsing the given snapshot with `JSON.parse` and setting it to `instanceData`.

I think 2 would be a good approach for now because we don't need an external library and it seems to fulfill the purpose of a snapshot well enough.

What do you think?

alxflam replied:
yes, that would be better. I applied the proposed stringify approach :+1:

lucas-koehler reviewed (commented):
Thank you for the update and the explanations! I guess the separation between `data` in the resource editor and `instanceData` in the base editor was a mistake or a leftover from past refactorings. Thanks for fixing that.

When using "save as" in the example, the save seems to work but I get an error message in the UI:
![image](https://user-images.githubusercontent.com/6959840/161998984-62ca8b3e-8a05-4ab4-887b-efd520583b97.png)
Quickly looking for this message in the Theia code, the reason seems to be a missing `applySnapshot` method. (see `copyAndSave` in `workspace-frontend-contribution`)

alxflam commented:
@lucas-koehler thanks for testing, i missed out checking the examples. Now the example should work. I've implemented `applySnapshot` for the `ResourceTreeEditorWidget` (notice that `setTreeData` needs to be called to ensure the view for the newly saved file is also updated and shows the applied state). The object type used for `createSnapshot` and `applySnapshot`is the same one the `MonacoEditorModel` uses (as we need some kind of object to wrap the primitive string).

lucas-koehler commented:
@alxflam Thanks for the update and the contribution in general :). The error message no longer appears :)
The CI currently fails but that does not seem to be related to your changes (it cannot download a dependency). I'll rerun it again in a few hours to see if it works better again :D
